tests/lean/byCasesMetaM.lean

```diff
@@ -6,8 +6,8 @@ open Lean in
 open Lean.Meta in
 def test : MetaM Unit := do
   let type := (← getConstInfo ``ex).type
-  let mvar ← mkFreshExprMVar type
-  let (#[p, q, h], mvarId) ← mvar.mvarId!.introNP 3 | throwError "unexpected"
+  let mv ← mkFreshExprMVar type
```

Comment:
This suggests to me that the linter should probably not fire on pattern matches with only one arm / alternative.

Reply:
I would tend to disagree here - the example code in the PR text can lead a new user without much experience to think that their function is complete, because there's no red squigglies to give the hint that a case was forgotten.

In any case, the updated version of the PR code now has a mechanism for excepting certain constructors from the lint, and it's going to be applied to `mvar`, among others, so we won't need to change this style if we don't want to.

Reply:
(I should revert these test changes in anticipation of that, I suppose)

Reply:
@david-christiansen 

> I would tend to disagree here - the example code in the PR text can lead a new user without much experience to think that their function is complete, because there's no red squigglies to give the hint that a case was forgotten.

In the `Tree` PR example, the unused variables linter fires, so there is still a hint that something is amiss. I would also argue that even a novice user would be unlikely to leave the `depth` function as is, because it still looks to produce a constant depth of `0`. On the flip side, this certainly may be less clear in more complex examples. However, I think the unused variables linter is likely still a good enough hint there.

> In any case, the updated version of the PR code now has a mechanism for excepting certain constructors from the lint, and it's going to be applied to `mvar`, among others, so we won't need to change this style if we don't want to.

Marking constructors seems like an incredibly invasive way to handle this problem. The `mvar`/`fvar` pattern of naming a variable in `let` after a constructor when the object in question is known to be that alternative seems like it will be common in user code . Needing to mark all of these constructors `allow_name_as_variable` seems burdensome (and possibly quite difficult if the type is defined in another package).

Reply:
Thanks for the feedback!

The unused variables linter isn't quite enough for situations where users are struggling. I've also seen this pattern, though not at the recent event (the examples/labs didn't include enumeration-looking things):
```
inductive Day where
  | monday
  | tuesday
  | wednesday

def num : Day → Nat
  | monday => 0
  | tuesday => 1
  | wednesday => 2
```
The last two patterns are marked "redundant alternative", and because the definition contains an error, the unused variables linter doesn't fire at all. This linter suggests a `.` before all three patterns, even in the presence of the error message, which can plausibly unstick those who are still struggling with the namespace system.

> Needing to mark all of these constructors allow_name_as_variable seems burdensome (and possibly quite difficult if the type is defined in another package).

Depending on the situation, it's also reasonable to `set_option linter.constructorNameAsVariable false`. All style linters have a cost - this one seems to benefit new users at the cost of experts, which seems like an OK prioritization to me.

In the Lean codebase, only a very few datatypes regularly use constructor names as variables, and newly-written code seems unlikely to have a hard time avoiding these lints if they don't want to set the option.


I think your concern could also be alleviated by having this linter only fire when the constructor has arity 0. In practice, it doesn't fire for constructors like `Tree.branch`, which instead say that the pattern is invalid: `invalid pattern, constructor or constant marked with '[match_pattern]' expected`. What if this linter was for arity-0 constructors only, and then that error message additionally checked for a similar constructor name and then made the same suggestion?

Reply:
@david-christiansen I agree that this linter should fire in that example. Note that my suggestion was only to not fire in matches with a single branch / arm and your example has 3. 😉 

>  What if this linter was for arity-0 constructors only, and then that error message additionally checked for a similar constructor name and then made the same suggestion?

I think this heuristic is a fine alternative solution that could cover most instances of the problem I raised as well, 😄 

---

Below are some additional comments  I had, but they are not really that important to progressing this issue, so feel free to skip reading them.

> Depending on the situation, it's also reasonable to `set_option linter.constructorNameAsVariable false`. All style linters have a cost - this one seems to benefit new users at the cost of experts

The ability to get around this with an option suggests to me that this is likely to be helpful for more advanced users to fix mistakes and/or disable it when they know there is none. Novice users are likely to get confused if/when this triggers in places were they probably don't even realize a pattern match is occurring. When I first saw this error case, I was confused as to why this linter was firing on a basic `let` declaration. Only after looking at it for while did it strike me that all `let` declarations are technically pattern matches. I imagine a novice user would be entirely bamboozled.

> In the Lean codebase, only a very few datatypes regularly use constructor names as variables, and newly-written code seems unlikely to have a hard time avoiding these lints if they don't want to set the option.

I really think you are underestimating how easy it is to stumble into this problem with a `let`. It is quite natural to write a `let` declaration with a constructor name for a type with smart constructors (e.g., `Expr`, `Syntax`, `Json`) For example, `let obj := Json.mkObj`  or `let atom := Syntax.mkAtom` would trigger the linter as-is. However, it could be argued that such names are not very good and thus are a code smell all on their own. However, I think that depends on the context. If the function is dealing with a single generic `obj`, `atom`, this can be quite natural to write and easy to understand.

Separately, one of the reasons I think these examples don't appear more in core outside of specialist types like `Expr` and `Syntax` is that there are surprisingly few complex inductive types in core. Structure-like inductives, abstract combinators (e.g., `Option`, `Except`, `Sum`) and arity 0 enumerations (e.g., `Ordering`, kinds, modes, flags, etc.) make up the lion's share of inductive types in core.

Reply:
This is useful, thanks for helping me refine it. I'll update the linter to fire on unary constructors only, and the error message above as well. I think that'll serve new users at least as well, with less of an imposition on experts.

Reply:
OK, looking at the error message in question, I don't think it's feasible to do type-based constructor name suggestions. I'll see if just listing out all the overloads of the name that would be eligible as patterns gives a good enough experience.

Reply:
I think it works pretty well, so long as not *all* the eligible overloads are listed!

Writing:
```
def length (list : List α) : Nat :=
  match list with
  | nil => 0
  | cons x xs => length xs + 1
```
now results in the following warning on `nil`:
```
warning: Local variable 'nil' resembles constructor 'List.nil' - write '.nil' (with a dot) or 'List.nil' to use the constructor.
note: this linter can be disabled with `set_option linter.constructorNameAsVariable false`
```

and error on `cons`:
```
invalid pattern, constructor or constant marked with '[match_pattern]' expected

Suggestion: 'List.cons' is similar
```

What do you think?

Reply:
Looks good to me!